lib/spack/spack/solver/counter.py

```diff
@@ -199,109 +199,106 @@ def possible_dependencies(
-                    continue
-            else:
+            pkg_cls = self.context.repo.get_pkg_class(pkg_name=pkg_name)
+            for name, conditions in pkg_cls.dependencies_by_name(when=True).items():
                 if all(
-                    dep.depflag != depflag for deplist in conditions.values() for dep in deplist
+                    self.context.unreachable(pkg_name=pkg_cls.name, when_spec=x)
+                    for x in conditions
                 ):
+                    tty.debug(
+                        f"[{__name__}] Not adding {name} as a dep of {pkg_cls.name}, because "
+                        f"conditions cannot be met"
+                    )
                     continue
 
-            if self.context.is_virtual(name) and name in virtuals:
-                continue
-
-            # expand virtuals if enabled, otherwise just stop at virtuals
-            if self.context.is_virtual(name):
-                virtuals.add(name)
-                if expand_virtuals:
-                    providers = self.context.providers_for(name)
-                    dep_names = {spec.name for spec in providers}
+                # check whether this dependency could be of the type asked for
+                if strict_depflag is False:
+                    depflag_union = 0
+                    for deplist in conditions.values():
+                        for dep in deplist:
+                            depflag_union |= dep.depflag
+                    if not (depflag & depflag_union):
+                        continue
                 else:
-                    visited.setdefault(pkg_cls.name, set()).add(name)
-                    visited.setdefault(name, set())
+                    if all(
+                        dep.depflag != depflag
+                        for deplist in conditions.values()
+                        for dep in deplist
+                    ):
+                        continue
+
+                if self.context.is_virtual(name) and name in virtuals:
                     continue
-            else:
-                dep_names = {name}
 
-            visited.setdefault(pkg_cls.name, set()).update(dep_names)
+                # expand virtuals if enabled, otherwise just stop at virtuals
+                if self.context.is_virtual(name):
+                    virtuals.add(name)
+                    if expand_virtuals:
+                        providers = self.context.providers_for(name)
+                        dep_names = {spec.name for spec in providers}
+                    else:
+                        visited.setdefault(pkg_cls.name, set()).add(name)
+                        visited.setdefault(name, set())
+                        continue
+                else:
+                    dep_names = {name}
 
-            for dep_name in dep_names:
-                if dep_name in visited:
-                    continue
+                visited.setdefault(pkg_cls.name, set()).update(dep_names)
```

Comment:
why is pkg_cls.name used here? is it different from pkg_name?

Reply:
No, this one was not caught by the refactoring tool

Reply:
Fixed in ee2c25500e9a72a81289c8aa04c2ff2883b05394